app/ui/client/lib/UserAction.js

```diff
@@ -0,0 +1,177 @@
+import { Meteor } from 'meteor/meteor';
+import { Tracker } from 'meteor/tracker';
+import { ReactiveDict } from 'meteor/reactive-dict';
+import { Session } from 'meteor/session';
+import _ from 'underscore';
+
+import { settings } from '../../../settings';
+import { Notifications } from '../../../notifications';
+
+const timeout = 15000;
+const renew = timeout / 3;
+const rooms = {};
+
+export const USER_RECORDING = 'user-recording';
+export const USER_TYPING = 'user-typing';
+export const USER_UPLOADING = 'user-uploading';
+export const USER_ACTIVITY = 'user-activity';
+
+const recordingTimeouts = {};
+const uploadingTimeouts = {};
+const typingTimeouts = {};
+
+const recordingRenews = {};
+const uploadingRenews = {};
+const typingRenews = {};
+
+const recordingUsers = new ReactiveDict();
+const uploadingUsers = new ReactiveDict();
+const typingUsers = new ReactiveDict();
+
+const selfActivity = { USER_RECORDING: false, USER_UPLOADING: false, USER_TYPING: false };
+
+const shownName = function(user) {
+	if (!user) {
+		return;
+	}
+	if (settings.get('UI_Use_Real_Name')) {
+		return user.name;
+	}
+	return user.username;
+};
+
+const stopActivity = (rid, activityType, extras) => Notifications.notifyRoom(rid, USER_ACTIVITY, shownName(Meteor.user()), false, activityType, extras);
+const startActivity = (rid, activityType, extras) => Notifications.notifyRoom(rid, USER_ACTIVITY, shownName(Meteor.user()), true, activityType, extras);
+
+function handleStreamAction(activity, performingUsers, rid, username, extras) {
+	const id = extras?.tmid ? extras.tmid : rid;
+	const users = performingUsers.get(id) || {};
```

Comment:
Here you can have a global `performingUsers` variable instead of 3 different dicts. And value could have all activities. Eg:

```
const activities = performingUsers.get(id) || {};
const users = activities[activityType];
```